### DISCUSSION TITLE
MVVM  구조 공부하기

### DISCUSSION THREAD

**Discussion by Jsueeee** [euphony-io/check-in-listener#5] [General]
@euphony-io/eu-us

저희 팀이 이번에 MVVM  구조를 다같이 열심히 공부해서 적용해보도록 결정했습니다.
아래에 참고하실만한 자료를 남깁니다.
다른 분들도 좋은 자료 있으시면 공유해주시고, 이해 안가는 건 여기서 같이 질문하면서 생각해봐요!

https://blog.yena.io/studynote/2019/03/16/Android-MVVM-AAC-1.html
https://blog.yena.io/studynote/2019/03/27/Android-MVVM-AAC-2.html

**Comment by yeonns** [comment]
Google CodeLab에 View Model과 Repository를 사용하는 실습예제가 있어서 공유드려요! :smile:  Room도 포함해서 둘다 도움이되실거같습니다 :)
[CodeLab Sample](https://developer.android.com/codelabs/android-room-with-a-view-kotlin?hl=ko&continue=https%3A%2F%2Fcodelabs.developers.google.com%2F#0)

**Reply by Jsueeee** [reply]
> Google CodeLab에 View Model과 Repository를 사용하는 실습예제가 있어서 공유드려요! 😄 Room도 포함해서 둘다 도움이되실거같습니다 :)
> [CodeLab Sample](https://developer.android.com/codelabs/android-room-with-a-view-kotlin?hl=ko&continue=https%3A%2F%2Fcodelabs.developers.google.com%2F#0)

@euphony-io/eu-us 모두 멘토님이 올려주신 링크 보셨나요??
지금은 insert(save)함수가 mainActivity에 있는데 ViewModel로 옮겨볼까요?

**Reply by sujin9** [reply]
네 main보다 옮기는게 더 좋을것같아요! 아니면 Repository 클래스를 만들어서 View Model에서는 DB에 직접적으로 접근 할 필요 없게 하는 방식으로 많이 하는 것 같던데, 이건 너무 번거로울까요?

**Reply by Jsueeee** [reply]
> 네 main보다 옮기는게 더 좋을것같아요! 아니면 Repository 클래스를 만들어서 View Model에서는 DB에 직접적으로 접근 할 필요 없게 하는 방식으로 많이 하는 것 같던데, 이건 너무 번거로울까요?

오 완전 좋은 것 같은데요???

**Reply by Jsueeee** [reply]
> 네 main보다 옮기는게 더 좋을것같아요! 아니면 Repository 클래스를 만들어서 View Model에서는 DB에 직접적으로 접근 할 필요 없게 하는 방식으로 많이 하는 것 같던데, 이건 너무 번거로울까요?

의견 반영해서 #30 에 레포지토리를 한 번 만들어봤습니다! 확인 부탁드려요~

**Comment by Jsueeee** [comment]
@euphony-io/eu-us 
회의 때 나왔던 권한 요청 관련해서 참고 자료 링크 남깁니다!
권한 관련 mvvm 적용은 이렇게 하면 간단하게 끝날 것 같습니다!
https://stackoverflow.com/questions/44550851/how-to-check-permission-is-granted-in-viewmodel

**Reply by Jsueeee** [reply]
액티비티에서는 라이브데이터로 권한이 허락되었는지 알 수 있게 만드는 게 제일 이상적인 것 같습니다.
첫 답변 참고하시면 좋을 것 같아요 :)

**Comment by Jsueeee** [comment]
# 구성요소


### Model

- 데이터 저장, 수정, 삭제 관리, 네트워크 통신 등
- Entitiy, Repository, DB, Util 클래스 등

### View

- UI
- Activity, Fragment, View, ViewDataBinding 등

### ViewModel

- View에 표현할 데이터를 Model로부터 가져와 가공하고 관리
- AAC ViewModel 클래스, BaseObservable 등

**Comment by sujin9** [comment]
파일 추출을 할 때 Live Data 적용을 해보려고 했는데 계속 null이 반환되고, observer를 사용하자니 LifecycleOwner 설정에 어려움이 있어서 
자료를 찾아보던 중 아래 자료들을 찾았는데, Live Data와 Repository 사용에 관하여 잘 얘기해주는 자료 같아서 공유해봅니다!

[링크1](https://medium.com/androiddevelopers/viewmodels-and-livedata-patterns-antipatterns-21efaef74a54) _(LiveData in repositories 부분 보시면 됩니다!)_ , [링크2](https://developer.android.com/topic/libraries/architecture/livedata#transform_livedata)

저도 LiveData & MVVM 모두 처음이라 지금도 계속 자료를 읽고 있는데요,
` 활동과 프래그먼트는 상태 보유가 아닌 데이터를 표시하는 역할을 하므로 LiveData 인스턴스를 보유해서는 안 됩니다. `
`데이터 영역 클래스에서 LiveData 객체를 작업하고 싶을 수 있지만 LiveData는 비동기 데이터 스트림을 처리하도록 설계되지 않았습니다.`
우선 예제들을 보면 제 생각에는 Repository에서는 Live Data 를 보유하지 않거나 asLiveData()를 사용해야 하는 것 같습니다!

그리고 ViewModel에서는
```kotlin
private fun getPostalCode(address: String): LiveData<String> {
        // DON'T DO THIS
        return repository.getPostCode(address)
    }
```
위와 같은 방법보다는 Transformation 사용을 권장하더라고요!

다른 분들도 한 번 보시고, 저희 repository랑 view model 코드에 수정이 필요할지, 
그리고 파일 추출에서 live data를 적용할 필요가 있을지 등에 대해 논의해봤으면 합니다!
